subsys/bluetooth/host/audio/iso.c

```diff
@@ -0,0 +1,1075 @@
+
+	/* check transitions validness */
+	switch (state) {
+	case BT_ISO_DISCONNECTED:
+		/* regardless of old state always allows this state */
+		break;
+	case BT_ISO_BOUND:
+		if (chan->state != BT_ISO_DISCONNECTED) {
+			BT_WARN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	case BT_ISO_CONNECT:
+		if (chan->state != BT_ISO_BOUND) {
+			BT_WARN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	case BT_ISO_CONNECTED:
+		if (chan->state != BT_ISO_BOUND &&
+		    chan->state != BT_ISO_CONNECT) {
+			BT_WARN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	case BT_ISO_DISCONNECT:
+		if (chan->state != BT_ISO_CONNECTED) {
+			BT_WARN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	default:
+		BT_ERR("%s()%d: unknown (%u) state was set", func, line, state);
+		return;
+	}
+
+	chan->state = state;
+}
+#else
+void bt_iso_chan_set_state(struct bt_iso_chan *chan, uint8_t state)
+{
+	chan->state = state;
+}
+#endif /* CONFIG_BT_AUDIO_DEBUG_ISO */
+
+void bt_iso_chan_remove(struct bt_conn *conn, struct bt_iso_chan *chan)
+{
+	struct bt_iso_chan *c;
+	sys_snode_t *prev = NULL;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&conn->channels, c, node) {
+		if (c == chan) {
+			sys_slist_remove(&conn->channels, prev, &chan->node);
+			return;
+		}
+
+		prev = &chan->node;
+	}
+}
+
+int bt_iso_chan_bind(struct bt_conn **conns, uint8_t num_conns,
+		     struct bt_iso_chan **chans)
+{
+	struct bt_iso_create_param param;
+	int i, err;
+
+	__ASSERT_NO_MSG(conns);
+	__ASSERT_NO_MSG(num_conns);
+	__ASSERT_NO_MSG(chans);
+
+	memset(&param, 0, sizeof(param));
+
+	/* TODO: Allow setting ID? */
+	param.id = 0x01;
+	param.num_conns = num_conns;
+	param.conns = conns;
+	param.chans = chans;
+
+	err = bt_conn_bind_iso(&param);
+	if (err) {
+		return err;
+	}
+
+	/* Bind respective connection to channel */
+	for (i = 0; i < num_conns; i++) {
+		bt_iso_chan_add(conns[i], chans[i]);
+		bt_iso_chan_set_state(chans[i], BT_ISO_BOUND);
+	}
+
+	return 0;
+}
+
+int bt_iso_chan_connect(struct bt_iso_chan **chans, uint8_t num_chans)
+{
+	struct bt_conn *conns[CONFIG_BT_MAX_ISO_CONN];
+	int i, err;
+
+	__ASSERT_NO_MSG(chans);
+	__ASSERT_NO_MSG(num_chans);
+
+	for (i = 0; i < num_chans; i++) {
+		if (!chans[i]->conn) {
+			return -ENOTCONN;
+		}
+
+		conns[i] = chans[i]->conn;
+	}
+
+	err = bt_conn_connect_iso(conns, num_chans);
+	if (err) {
+		return err;
+	}
+
+	for (i = 0; i < num_chans; i++) {
+		bt_iso_chan_set_state(chans[i], BT_ISO_CONNECT);
+	}
+
+	return 0;
+}
+
+int bt_iso_chan_disconnect(struct bt_iso_chan *chan)
+{
+	__ASSERT_NO_MSG(chan);
+
+	if (!chan->conn) {
+		return -ENOTCONN;
+	}
+
+	if (chan->state == BT_ISO_BOUND) {
+		bt_iso_chan_set_state(chan, BT_ISO_DISCONNECTED);
+		bt_conn_unref(chan->conn);
+		chan->conn = NULL;
+		return 0;
+	}
+
+	return bt_conn_disconnect(chan->conn, BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+}
+
+void bt_iso_recv(struct bt_conn *conn, struct net_buf *buf, uint8_t flags)
+{
+	struct bt_hci_iso_data_hdr *hdr;
```

Comment:
Do we need to check or handle the `sn` in this function? Seems to be largely ignored.

Reply:
Good point, I will need to check if the sn is there to actually tell for which packet it is, though I don't think we can get multiple outstanding packets being reassembled so I guess we could just compare if the sn matches or not, and if doesn't discard the all fragments.

Reply:
Agreed